HydrateOrDiedrate/src/Piping/ShutoffValve/BlockShutoffValve.cs

```diff
@@ -0,0 +1,270 @@
+using System;
+using System.Collections.Generic;
+using HydrateOrDiedrate.Piping.FluidNetwork;
+using HydrateOrDiedrate.Wells.WellWater;
+using Vintagestory.API.Common;
+using Vintagestory.API.MathTools;
+using Vintagestory.API.Util;
+
+namespace HydrateOrDiedrate.Piping.ShutoffValve
+{
+    public enum ValveAxis { EW, NS, UD }
+
+    public class BlockShutoffValve : Block, IFluidBlock, IFluidGate
+    {
+        static readonly (ValveAxis axis, BlockFacing a, BlockFacing b)[] ScanOrder = new[]
+        {
+            (ValveAxis.EW, BlockFacing.EAST,  BlockFacing.WEST),
+            (ValveAxis.NS, BlockFacing.NORTH, BlockFacing.SOUTH),
+            (ValveAxis.UD, BlockFacing.UP,    BlockFacing.DOWN)
+        };
+        static readonly BlockFacing[] Faces =
+        {
+            BlockFacing.NORTH, BlockFacing.EAST, BlockFacing.SOUTH,
+            BlockFacing.WEST,  BlockFacing.UP,   BlockFacing.DOWN
+        };
+
+        void MarkSelfAndNeighborsDirty(IWorldAccessor world, BlockPos pos)
+        {
+            var ba = world.BlockAccessor;
+            ba.MarkBlockDirty(pos);
+            foreach (var f in Faces)
+            {
+                ba.MarkBlockDirty(pos.AddCopy(f));
+            }
+        }
+
+        public bool HasFluidConnectorAt(IWorldAccessor world, BlockPos pos, BlockFacing face)
+        {
+            var be = world.BlockAccessor.GetBlockEntity(pos) as BlockEntityShutoffValve;
+            if (be == null) return false;
+            switch (be.Axis)
+            {
+                case ValveAxis.EW: return face == BlockFacing.EAST || face == BlockFacing.WEST;
+                case ValveAxis.NS: return face == BlockFacing.NORTH || face == BlockFacing.SOUTH;
+                default:           return face == BlockFacing.UP   || face == BlockFacing.DOWN;
+            }
+        }
+
+        public bool AllowsFluidPassage(IWorldAccessor world, BlockPos pos, BlockFacing from, BlockFacing to)
+        {
+            var be = world.BlockAccessor.GetBlockEntity(pos) as BlockEntityShutoffValve;
+            if (be == null) return false;
+            if (to != from.Opposite) return false;
+            bool alongAxis = be.Axis switch
+            {
+                ValveAxis.EW => (from == BlockFacing.EAST  && to == BlockFacing.WEST) ||
+                                (from == BlockFacing.WEST  && to == BlockFacing.EAST),
+                ValveAxis.NS => (from == BlockFacing.NORTH && to == BlockFacing.SOUTH) ||
+                                (from == BlockFacing.SOUTH && to == BlockFacing.NORTH),
+                ValveAxis.UD => (from == BlockFacing.UP    && to == BlockFacing.DOWN) ||
+                                (from == BlockFacing.DOWN  && to == BlockFacing.UP),
+                _ => false
+            };
```

Comment:
You already confirmed that `to` and `from` are opposite so no need to check them both.

```suggestion
            bool alongAxis = be.Axis switch
            {
                ValveAxis.EW => to == BlockFacing.EAST  || to == BlockFacing.WEST,
                ValveAxis.NS => to == BlockFacing.NORTH || to == BlockFacing.SOUTH,
                ValveAxis.UD => to == BlockFacing.UP    || to == BlockFacing.DOWN,
                _ => false
            };
```